core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -511,17 +617,27 @@ private boolean isCorrelated(Expression expression)
 
         // mapping from correlated symbols to their uncorrelated equivalence
         final Multimap<Symbol, Symbol> correlatedSymbolsMapping;
+
+        // local (uncorrelated) symbols known to be constant based on their dependency on correlation symbols
+        // they are derived from the filter predicate, e.g.
+        // a = corr --> a is constant
+        // b = f(corr_1, corr_2, ...) --> b is constant provided that f is deterministic
+        // cast(c AS ...) = corr --> c is constant provided that cast is injective
```

Comment:
I wonder if this case could be handled by expanding the UnwrapCastInComparison optimizer to deal with correlation symbols on the right hand side. In that case, the expression would be turned to something like:

```
c = cast(corr as ...)
```

and the second rule would apply.

That way, we don't need the decorrelator to deal with cast and checking whether it's injective at all.

Reply:
The transformation which is done in `UnwrapCastInComparison` is based on knowing the actual value of the right-hand-side expression. Here, we only know that it's constant. If apply the cast to `corr`, it might turn out to be a truncating cast, and the result of the comparison might change. Or we could get a cast exception.

Reply:
Fair enough. In the future, we might want to add an optimization that works when the right side is constant -- it would be a weaker form of UnwrapCastInComparison.